imap/guesstz.c

```diff
@@ -0,0 +1,914 @@
+            seconds < 0 || seconds > 60) {
+        return 0;
+    }
+
+    snprintf(buffer, 8, "%c%02i%02i", sign, hours, minutes);
+    if (seconds) {
+        snprintf(buffer + 5, 3, "%02i", seconds);
+    }
+
+    return strlen(buffer);
+}
+
+
+static void expand_observances(icalcomponent *vtz,
+                       icaltimetype start, icaltimetype end,
+                       dynarray_t *obsrecs)
+{
+    const icaltimezone *utc = icaltimezone_get_utc_timezone();
+    icalarray *obsarray;
+
+
+    icalcomponent *myvtz = icalcomponent_clone(vtz);
+    obsarray = icalarray_new(sizeof(struct observance), 20);
+    icaltimezone_truncate_vtimezone_advanced(myvtz, &start, &end, obsarray,
+            NULL, NULL, NULL, NULL, 0);
+    icalcomponent_free(myvtz);
+
+    if (!obsarray->num_elements) {
+        icalarray_free(obsarray);
+        return;
+    }
+
+    unsigned i;
+    for (i = 0; i < obsarray->num_elements; i++) {
+        struct observance *obs = icalarray_element_at(obsarray, i);
+        struct obsrec obsrec = {
+            (int64_t) icaltime_as_timet_with_zone(obs->onset, utc),
+            (int32_t) obs->offset_to
+        };
+        dynarray_append(obsrecs, &obsrec);
+    }
+
+    icalarray_free(obsarray);
+}
+
+static int offsetrec_cmp(const void *va, const void *vb)
+{
+    const struct offsetrec *a = va;
+    const struct offsetrec *b = vb;
+
+    if (a->offset != b->offset) {
+        return a->offset < b->offset ? -1 : 1;
+    }
+    if (a->tznum != b->tznum) {
+        return a->tznum < b->tznum ? -1 : 1;
+    }
+
+    return 0;
+}
+
+static char *guesstz_fname(const char *zoneinfo_dir, const char *alt_fname)
+{
+    struct buf fname = BUF_INITIALIZER;
+
+    if (!zoneinfo_dir) {
+        zoneinfo_dir = config_getstring(IMAPOPT_ZONEINFO_DIR);
+        if (!zoneinfo_dir) {
+            xsyslog(LOG_ERR, "No zoneinfo_dir found in config", NULL);
+            return NULL;
+        }
+    }
+    if (!alt_fname) {
+        buf_setcstr(&fname, zoneinfo_dir);
+        if (fname.s[fname.len-1] != '/') buf_putc(&fname, '/');
+        buf_appendcstr(&fname, "guesstz.db");
+    }
+    else {
+        buf_setcstr(&fname, alt_fname);
+    }
+    buf_cstring(&fname);
+
+    return buf_release(&fname);
+}
+
+static int store_timezone(struct guesstzdb *gtzdb,
+                          uint32_t tznum, const char *tzid,
+                          const struct obsrec *obs, size_t nobs)
+{
+    struct buf *key = &gtzdb->buf1;
+    buf_reset(key);
+    buf_putc(key, 'T');
+    buf_appendbit32(key, tznum);
+
+    struct buf *data = &gtzdb->buf2;
+    buf_setcstr(data, tzid);
+    buf_putc(data, '\0');
+
+    size_t tzidlen = buf_len(data);
+    buf_truncate(data, tzidlen + + nobs * sizeof(struct obsrec));
+    struct obsrec *myobs = (struct obsrec *) (data->s + tzidlen);
+    size_t i;
+    for (i = 0; i < nobs; i++) {
+        myobs[i].onset = htonll(obs[i].onset);
+        myobs[i].offset = htonl(obs[i].offset);
+    }
```

Comment:
I understand what's going on here (putting ints into a char array) but I wonder if we couldn't use buf_appendbit32() and buf_appendbit64()

Reply:
That's the better approach, indeed. The current approach suffered from struct padding anyway.